libpoolprotocols/stratum/EthStratumClient.cpp

```diff
@@ -411,131 +420,423 @@ void EthStratumClient::processReponse(Json::Value& responseObject)
-	{
+	// Handle awaited responses to OUR requests
+	if (!_isNotification) {
+
+		Json::Value jReq;
+		Json::Value jPrm;
+
+		switch (_id)
+		{
+
 		case 1:
 
 			// Response to "mining.subscribe"
 			switch (m_conn.Version()) {
 
 			case EthStratumClient::STRATUM:
 
-				m_subscribed.store(responseObject.get("result", Json::Value::null).asBool(), std::memory_order_relaxed);
+				m_subscribed.store(_isSuccess, std::memory_order_relaxed);
 				if (!m_subscribed)
 				{
 					cnote << "Could not subscribe to stratum server";
 					disconnect();
 					return;
 				}
+				else {
+
+					cnote << "Subscribed to stratum server";
+
+					jReq["id"] = unsigned(3);
+					jReq["jsonrpc"] = "2.0";
+					jReq["method"] = "mining.authorize";
+					jReq["params"] = Json::Value(Json::arrayValue);
+					jReq["params"].append(m_conn.User());
+					jReq["params"].append(m_conn.Pass());
+
+				}
 
-				cnote << "Subscribed to stratum server";
-				sendSocketData("{\"id\": 3, \"method\": \"mining.authorize\", \"params\": [\"" + m_conn.User() + "\",\"" + m_conn.Pass() + "\"]}");
 				break;
 
 			case EthStratumClient::ETHPROXY:
 
-				sendSocketData("{\"id\": 5, \"method\": \"eth_getWork\", \"params\": []}"); // not strictly required but it does speed up initialization
+				m_subscribed.store(_isSuccess, std::memory_order_relaxed);
+				if (!m_subscribed)
+				{
+					cnote << "Could not login to ethproxy server";
+					disconnect();
+					return;
+				}
+				else {
+
+					cnote << "Logged in to eth-proxy server";
+					m_authorized.store(true, std::memory_order_relaxed);
+
+					jReq["id"] = unsigned(5);
+					jReq["method"] = "eth_getWork";
+					jReq["params"] = Json::Value(Json::arrayValue);
+
+				}
+
 				break;
 
 			case EthStratumClient::ETHEREUMSTRATUM:
 
-				m_nextWorkDifficulty = 1;
-				params = responseObject.get("result", Json::Value::null);
-				if (params.isArray())
+				m_subscribed.store(_isSuccess, std::memory_order_relaxed);
+				if (!m_subscribed)
 				{
-					std::string enonce = params.get((Json::Value::ArrayIndex)1, "").asString();
-					processExtranonce(enonce);
+					cnote << "Could not subscribe to stratum server";
+					disconnect();
+					return;
 				}
+				else {
+
+					cnote << "Subscribed to stratum server";
+
+					m_nextWorkDifficulty = 1;
+					jPrm = responseObject.get(((_rpcVer==1) ? "result":"params"), Json::Value::null);
+
+					if (!jPrm.empty() && jPrm.isArray()) {
+						std::string enonce = jPrm.get((Json::Value::ArrayIndex)1, "").asString();
+						processExtranonce(enonce);
+					}
+
+					jReq["id"] = unsigned(2);
+					jReq["method"] = "mining.extranonce.subscribe";
+					jReq["params"] = Json::Value(Json::arrayValue);
+
+				}
+
 
-				sendSocketData("{\"id\": 2, \"method\": \"mining.extranonce.subscribe\", \"params\": []}");
 				break;
 			}
 
-		break;
+			sendSocketData(jReq);
+			break;
 
-	case 2:
+		case 2:
 
-		// nothing to do...
-		break;
+			// ETHEREUMSTRATUM *after* first mining.extranonce.subscribe
```

Comment:
why?
Can't we just send both (id2 and id3) at once or send auth (id3) first and once we get authorized id2 later?
As far as i understood the json rpc specs, what value we send as id does not matter its only to map request to response.

Reply:
I just modified the logic. id2 and 3 are sent in sequence as the response to id2 should not perform any logic.
Accroding to this:
https://github.com/nicehash/Specifications/blob/master/NiceHash_extranonce_subscribe_extension.txt